### ISSUE TITLE
USWDS - Bug: Consider using the <dialog> element for the Alert component

### ISSUE THREAD

**Issue by jenstrickland** [uswds/uswds#5993]
### Describe the bug

The alert component as it is designed doesn't convey any semantics without the use of ARIA. For non-sighted users they're unlikely to reach the component unless they are reading the whole screen or enough heading levels. There is an option to use an ARIA role, as noted under Alert ARIA roles, but this requires the use of ARIA, and…

Since the [first rule of ARIA is to use HTML and not ARIA](https://www.a11y-collective.com/blog/the-first-rule-for-using-aria/), please consider using the [`<dialog>` element.](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/dialog) There are a variety of ways to implement this, some of which can also help with your desire for a dismissible alert.

The component as it is currently designed wouldn't pass [WCAG 2.0 1.3.1 (Level A) ](https://www.w3.org/TR/UNDERSTANDING-WCAG20/content-structure-separation-programmatic.html) because unless the ARIA role is added it doesn't convey the relationship of the information programmatically. It's obvious that it is an alert or warning from the visual design but for those relying on assistive technology or no visual design, the alert status should be conveyed. The HTML dialog element was intended for such use cases. Until recently there wasn't full support, but [now it is safe to use `<dialog>`.](https://caniuse.com/dialog)

### Steps to reproduce the bug

1. Go to https://designsystem.digital.gov/components/alert/
2. Open up Component code
3. Note that the `<div>` element conveys no semantic context.
4. The `role="alert"` is added to the error and emergency headings, but not to the warning or other context. 

### Expected Behavior

Adding the `<dialog>` element semantics would programmatically communicate the context conveyed in the visual design. As [MDN states](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/dialog):
> The HTML <dialog> element is used to create both modal and non-modal dialog boxes. Modal dialog boxes interrupt interaction with the rest of the page being inert, while non-modal dialog boxes allow interaction with the rest of the page.

### System setup

- USWDS version: 3.8.1 - based on the component shown https://designsystem.digital.gov/components/alert/

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://designsystem.digital.gov/about/community/#community-conduct).
- [X] I checked the [current issues](https://github.com/uswds/uswds/issues?q=is%3Aissue+is%3Aopen+label%3A%22Type%3A+Bug%22) for duplicate bug reports.

**Comment by brunerae** [comment]
Hi @alex-hull / @amycole501 - wondering if one of you would be able to confirm this bug? Thanks!

**Comment by amycole501** [comment]
Having used the alert component before I can agree that the use of ARIA can be problematic especially on forms or other interactive pages where the alert may be overlooked by AT users.

I would love to test a version of the alert component using the dialog suggestion.

**Comment by jenstrickland** [comment]
If it's useful I can discuss with the IRS design system team and quickly mock up a few examples.

**Comment by amycole501** [comment]
@jenstrickland if your team has time to help with that I'd love to test it with screen readers!

**Comment by jenstrickland** [comment]
@amycole501 I proposed with my team, we're waiting to confirm it's okay. Soon!

**Comment by amycole501** [comment]
@jenstrickland Checking in on this, has your team been able to make any progress?

**Comment by jenstrickland** [comment]
Yes! I have prototypes to share. I started a response last week, then got pulled onto something else. Draft in progress, though!

**Comment by jenstrickland** [comment]
_Finally_, I have coded prototypes for this series of alerts to make them more semantic and improve the user experience for all, especially for assistive technology users. I added my HTML, CSS, & JS to a new repository: https://github.com/jenstrickland/20250106alerts

All the code is for demo purposes, and will need adjustments, but I did try to keep close to USWDS.

## Alert Dialog HTML Prototypes

- Informative status alerts: can include information, success, warning, or error, and persists on screen
- Warning status: needs acknowledgement to dismiss
- Emergency status: needs to contain attention and direct to an action

<hr/>

### Informative Status

- This could also be used for success or warning messages, _if_ they don't need to be acknowledged.

- Instead of a div + h4, recommend using a dl + dt, so that it can be inserted into any semantic structure without requiring a preceding h3. The `dt` can be enhanced with ARIA to be treated like a heading and use the proper `level` as well.

```
                        <dl class="usa-alert usa-alert--info">
                          <span class="usa-alert__body">
                            <dt class="usa-alert__heading">
                                Informative status
                            </dt>
                            <dd>
                                <p class="usa-alert__text">
                                  Lorem ipsum dolor sit amet,
                                <a class="usa-link" href="javascript:void(0);">consectetur adipiscing</a>
                                elit, sed do eiusmod.
                                </p>
                            </dd>
                          </span>
                        </dl>

                        <dl class="usa-alert usa-alert--success">
                          <span class="usa-alert__body">
                            <dt class="usa-alert__heading">
                                Success status
                            </dt>
                            <dd>
                                <p class="usa-alert__text">
                                  Lorem ipsum dolor sit amet,
                                <a class="usa-link" href="javascript:void(0);">consectetur adipiscing</a>
                                elit, sed do eiusmod.
                                </p>
                            </dd>
                          </span>
                        </dl>

                        <dl class="usa-alert usa-alert--error">
                          <span class="usa-alert__body">
                            <dt class="usa-alert__heading">
                                Error status
                            </dt>
                            <dd>
                                <p class="usa-alert__text">
                                  Lorem ipsum dolor sit amet,
                                <a class="usa-link" href="javascript:void(0);">consectetur adipiscing</a>
                                elit, sed do eiusmod.
                                </p>
                            </dd>
                          </span>
                        </dl>
```


<img width="919" alt="screenshot of informative alerts prototype" src="https://github.com/user-attachments/assets/693193b8-174a-4371-8c58-896841c7cc51" />


<hr/>

### Warning Status

- If the alert should be acknowledged and then dismissed, but content can still be navigated around it.

- The heading level needs to be dynamically set based on the surrounding content. In this case, presuming a preceding heading level is h1, this is an h2.

- Consider, how urgent is this warning? Does it need to be `role="alert"`?

```
<dialog autofocus open class="usa-alert usa-alert--warning">
  <span class="usa-alert__body">
    <h2 class="usa-alert__heading">Warning status</h2>
    <p class="usa-alert__text">
      Lorem ipsum dolor sit amet,
      <a class="usa-link" href="javascript:void(0);">consectetur adipiscing</a> elit, sed do eiusmod.
    </p>

    <form method="dialog" style="padding-top: .5rem">
      <button class="usa-button">OK</button>
    </form>
  </span>
</dialog>
```

<img width="891" alt="screenshot of alert warning prototype" src="https://github.com/user-attachments/assets/7bd00808-009c-4420-9b25-556fa566e943" />

<hr/>

### Emergency Status

The emergency message should be system-activated, use a modal and the ::backdrop CSS pseudo-element, trap focus within the dialog, and require an action.

```
<dialog role="alertdialog" autofocus aria-labelledby class="usa-alert usa-alert--emergency dialog">
  <div class="usa-alert__body dialog-overlay">
    <h2 class="usa-alert__heading" id="emergency">Emergency status</h2>
    <p class="usa-alert__text">
      Recommend to use backdrop CSS pseudo-element and contain/trap focus within modal. Lorem ipsum dolor sit amet,
      <a class="usa-link" href="javascript:void(0);">consectetur adipiscing</a> elit, sed do eiusmod.
    </p>
  </div>
  <button class="usa-button close">Close</button>
</dialog>
```

<img width="894" alt="screenshot of alert emergency overlay prototype" src="https://github.com/user-attachments/assets/74f33213-cb6a-4fc6-895d-115ac71f8ce3" />

<hr/>

I hope this is clear. It was a bit tricky to figure out how to put it all together. Naturally, let me know if I can clarify anything.

**Comment by amycole501** [comment]
I'm adding this to discuss internally and establish next steps. Thank you! We will reach out if we need any clarification.

**Comment by ryparker** [comment]
I'm bringing #6407 into this issue so we can discuss the alert problem, prototypes, and test results together.

@jenstrickland's [prototype](https://github.com/jenstrickland/20250106alerts) and [original explanation](https://github.com/uswds/uswds/issues/5993#issuecomment-2573783592) are the starting point. As the contributor noted, this is demonstration code that will need adjustment.

We still need a Storybook example or a focused test page comparing the existing USWDS alert with the proposed alternatives. Passive information and status messages, warnings that require acknowledgement, and urgent dialogs need separate testing. For each, we should record keyboard behavior, focus entry and return, whether people can find and hear the message, how dismissal works, and whether the rest of the page remains usable.

This also relates to the dismissible-alert research in #5754. We haven't decided to add a dismissible variant or replace every alert with `<dialog>`. Any change during the current maintenance period must continue to support existing markup and styling, and testing needs to show both the problem and a compatible fix before we choose a production change.

Closing #6407 only brings the tracking into one place. The accessibility testing and the original concern remain unresolved. I've kept its description below as a reference; it doesn't imply that the tests passed or that every proposed approach has been approved.

<details>
<summary>#6407: USWDS - Feature: Create prototype for using dialog element in USWDS alert</summary>

Source: https://github.com/uswds/uswds/issues/6407

### Is your feature request related to a problem? Please describe.

The a11y team needs a USWDS dev to build a dialog alert prototype using code from [this repo](https://github.com/jenstrickland/20250106alerts)

### Describe the solution you'd like

Create a testing environment in storybook or codepen so the a11y team can test the functionality of our alert component using the dialog element instead of the current ARIA alert roles.

### Describe alternatives you've considered

_No response_

### Additional context

Our aim is to evaluate the current alert accessibility against the dialog alert option. Our findings from this prototype test will be compiled into our research for [dismissible alerts](https://github.com/uswds/uswds/issues/5754).

### Code of Conduct

- [x] I agree to follow this project's [Code of Conduct](https://designsystem.digital.gov/about/community/#community-conduct).
- [x] I checked the [current issues](https://github.com/uswds/uswds/issues?q=is%3Aissue+is%3Aopen+label%3A%22Type%3A+Feature+Request%22) for duplicate feature requests.

</details>
